README.md
chipKIT DTWI-based TLC59116 LED driver


debian/changelog
tlc59116-dtwi (1.0.2) uecide; urgency=low

  * Fixed package name

 -- Matt Jenkins <matt@majenko.co.uk>  Thu, 21 Sep 2017 00:15:11 +0100

tlc59116-dtwi (1.0.1) uecide; urgency=low

  * Added build files

 -- Matt Jenkins <matt@majenko.co.uk>  Thu, 21 Sep 2017 00:14:31 +0100

tlc59116-dtwi (1.0.0) uecide; urgency=low

  * Initial Import

 -- Matt Jenkins <matt@majenko.co.uk>  Thu, 21 Sep 2017 00:12:09 +0100



TLC59116_DTWI.h
/*
 * Copyright (c) 2017, Majenko Technologies
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification, 
 * are permitted provided that the following conditions are met:
 * 
 *  1. Redistributions of source code must retain the above copyright notice, 
 *     this list of conditions and the following disclaimer.
 * 
 *  2. Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *      and/or other materials provided with the distribution.
 * 
 *  3. Neither the name of Majenko Technologies nor the names of its contributors may be used
 *     to endorse or promote products derived from this software without 
 *     specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE 
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL 
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, 
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


#ifndef _TLC59116_H
#define _TLC59116_H

#if (ARDUINO >= 100) 
# include <Arduino.h>
#else
# include <WProgram.h>
#endif

#include <DTWI.h>

#define TLC59116_BASEADDR   0b1100000
#define TLC59116_ALLCALL    0b1101000
#define TLC59116_SUB1       0b1101001
#define TLC59116_SUB2       0b1101010
#define TLC59116_SUB3       0b1101100
#define TLC59116_RESET      0b1101011

#define TLC59116_MODE1      0x00
#define TLC59116_MODE2      0x01
#define TLC59116_PWM0       0x02
#define TLC59116_PWM1       0x03
#define TLC59116_PWM2       0x04
#define TLC59116_PWM3       0x05
#define TLC59116_PWM4       0x06
#define TLC59116_PWM5       0x07
#define TLC59116_PWM6       0x08
#define TLC59116_PWM7       0x09
#define TLC59116_PWM8       0x0A
#define TLC59116_PWM9       0x0B
#define TLC59116_PWM10      0x0C
#define TLC59116_PWM11      0x0D
#define TLC59116_PWM12      0x0E
#define TLC59116_PWM13      0x0F
#define TLC59116_PWM14      0x10
#define TLC59116_PWM15      0x11
#define TLC59116_GRPPWM     0x12
#define TLC59116_GRPFREQ    0x13
#define TLC59116_LEDOUT0    0x14
#define TLC59116_LEDOUT1    0x15
#define TLC59116_LEDOUT2    0x16
#define TLC59116_LEDOUT3    0x17
#define TLC59116_SUBADR1    0x18
#define TLC59116_SUBADR2    0x19
#define TLC59116_SUBADR3    0x1A
#define TLC59116_ALLCALLADR 0x1B
#define TLC59116_IREF       0x1C
#define TLC59116_EFLAG1     0x1D
#define TLC59116_EFLAG2     0x1E

/*! The TLC59116 library drives a TLC59116 PWM LED Driver chip over I2C.  It allows the control of each output via
 *  functions that work the same as the normal Arduinop analogWrite() function.  There is also a pin mapping facility
 *  for displaying numeric values on a 7 segment display.
 */

class TLC59116_DTWI {
    private:
        uint8_t _addr;
        uint8_t _begun;
        DTWI *_dtwi;
        uint8_t _dp;
        bool _leading;

        void writeRegister(uint8_t reg, uint8_t val);
        static const uint8_t pinmap[8];
        static const uint8_t numbers[10];

        const uint8_t *_currentPinMapping;

    public:
        /*! Create a new TLC59116 board with the default address of 0 */
        TLC59116_DTWI(DTWI &dtwi) : _addr(0), _begun(0), _dtwi(&dtwi), _dp(0), _leading(false) {}
        TLC59116_DTWI(DTWI *dtwi) : _addr(0), _begun(0), _dtwi(dtwi), _dp(0), _leading(false) {}
        TLC59116_DTWI(DTWI &dtwi, uint8_t addr) : _addr(addr), _begun(0), _dtwi(&dtwi), _dp(0), _leading(false) {}
        TLC59116_DTWI(DTWI *dtwi, uint8_t addr) : _addr(addr), _begun(0), _dtwi(dtwi), _dp(0), _leading(false) {}
        /*! Initialize the board, set all channels to 0, and apply the default pin mapping */
        void begin();
        /*! Set the channel (0-15) to the given brightness value (0-255) */
        void analogWrite(uint8_t channel, uint8_t brightness);
        /*! Display the given number at the specified brightness using the current pin mapping */
        void displayNumber(uint8_t number, uint8_t brightness);
        /*! Load a new pin mapping.
         *
         * Pin mappings are in the form of an 8-byte array with each byte representing the mapping for segments A to G and DP in that order.
         * The upper nibble (4 bits) of each byte is the analog channel number for the left-hand digit. The lower nibble is the channel number for the right-hand digit.
         */
        void setPinMapping(const uint8_t *mapping);
        void setDecimalPoint(uint8_t dp) { _dp = dp & 0x03; }
        void setLeadingZero(bool l) { _leading = l; }
};
#endif


TLC59116_DTWI.cpp
/*
 * Copyright (c) 2017, Majenko Technologies
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification, 
 * are permitted provided that the following conditions are met:
 * 
 *  1. Redistributions of source code must retain the above copyright notice, 
 *     this list of conditions and the following disclaimer.
 * 
 *  2. Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *      and/or other materials provided with the distribution.
 * 
 *  3. Neither the name of Majenko Technologies nor the names of its contributors may be used
 *     to endorse or promote products derived from this software without 
 *     specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE 
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL 
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, 
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


#include <TLC59116_DTWI.h>

const uint8_t TLC59116_DTWI::pinmap[8] = {
    10 << 4 | 14,
    11 << 4 | 15,
    5 << 4 | 1,
    6 << 4 | 2,
    7 << 4 | 3,
    9 << 4 | 13,
    8 << 4 | 12,
    4 << 4 | 0
};

const uint8_t TLC59116_DTWI::numbers[] = {
    0b00111111,
    0b00000110,
    0b01011011,
    0b01001111,
    0b01100110,
    0b01101101,
    0b01111101,
    0b00100111,
    0b01111111,
    0b01101111
};

void TLC59116_DTWI::begin() {
    if (_begun == 0) {
        _dtwi->beginMaster(DTWI::FQ1MHz);
        writeRegister(TLC59116_MODE1, 0x01);
        delay(1);
        writeRegister(TLC59116_MODE2, 0x00);
        writeRegister(TLC59116_LEDOUT0, 0b10101010);
        writeRegister(TLC59116_LEDOUT1, 0b10101010);
        writeRegister(TLC59116_LEDOUT2, 0b10101010);
        writeRegister(TLC59116_LEDOUT3, 0b10101010);
        this->analogWrite(0, 0);
        this->analogWrite(1, 0);
        this->analogWrite(2, 0);
        this->analogWrite(3, 0);
        this->analogWrite(4, 0);
        this->analogWrite(5, 0);
        this->analogWrite(6, 0);
        this->analogWrite(7, 0);
        this->analogWrite(8, 0);
        this->analogWrite(9, 0);
        this->analogWrite(10, 0);
        this->analogWrite(11, 0);
        this->analogWrite(12, 0);
        this->analogWrite(13, 0);
        this->analogWrite(14, 0);
        this->analogWrite(15, 0);
        setPinMapping(pinmap);
    }
    _begun = 1;
}


void TLC59116_DTWI::writeRegister(uint8_t reg, uint8_t val) {
    uint8_t state = 0;
    uint32_t ts = millis();
    while (1) {
        if (millis() - ts > 100) {
            _dtwi->stopMaster();
            return;
        }
        switch (state) {
            case 0: // begin write 
                if (_dtwi->startMasterWrite(TLC59116_BASEADDR | (_addr & 0x0F))) {
                    state = 1;
                }
                break;
            case 1: // Send register address
                if (_dtwi->write(&reg, 1) == 1) {
                    state = 2;
                }
                break;
            case 2: // Send register address
                if (_dtwi->write(&val, 1) == 1) {
                    state = 3;
                }
                break;
            case 3: // Stop Master
                if (_dtwi->stopMaster()) {
                    return;
                }
                break;
        }
    }
}

void TLC59116_DTWI::analogWrite(uint8_t chan, uint8_t b) {
    writeRegister(TLC59116_PWM0 + (chan & 0x0F), b);
}

void TLC59116_DTWI::displayNumber(uint8_t number, uint8_t b) {

    uint8_t tens = (number / 10) % 10;
    uint8_t units = number % 10;

    if (tens == 0) {
        for (int i = 0; i < 8; i++) {
            int seg = _currentPinMapping[i] >> 4;
            this->analogWrite(seg, 0);
        }
    } else {
        int num = numbers[tens];
        for (int i = 0; i < 8; i++) {
            int seg = _currentPinMapping[i] >> 4;
            if (num & (1<<i)) {
                this->analogWrite(seg, b);
            } else {
                this->analogWrite(seg, 0);
            }
        }
    }

    int num = numbers[units];
    for (int i = 0; i < 8; i++) {
        int seg = _currentPinMapping[i] & 0x0F;
        if (num & (1<<i)) {
            this->analogWrite(seg, b);
        } else {
            this->analogWrite(seg, 0);
        }
    }

}

void TLC59116_DTWI::setPinMapping(const uint8_t *map) {
    _currentPinMapping = map;
}
